numpy/core/src/multiarray/compiled_base.c.src

```diff
@@ -8,38 +8,74 @@
 #include "numpy/npy_3kcompat.h"
 #include "numpy/npy_math.h"
 #include "npy_config.h"
+#include "numpy/halffloat.h"
 #include "templ_common.h" /* for npy_mul_with_overflow_intp */
 #include "lowlevel_strided_loops.h" /* for npy_bswap8 */
 #include "alloc.h"
 #include "common.h"
 
 
+static char
+select_most_suitable_float_type(char a, char b) {
+    int i = 0, a_i = 0, b_i = 0;
+
+    static char float_type_list[] = {'E', 'F', 'D', 'G'};
+
+    a = toupper(a);
+    b = toupper(b);
+
+    if (!('D' <= a && a <= 'G')) {
+        a = 'D';
+    }
+    if (!('D' <= b && b <= 'G')) {
+        b = 'D';
+    }
+
+    for (i = 0; i < 4; i++) {
+        if (a == float_type_list[i]) {
+            a_i = i;
+        }
+        if (b == float_type_list[i]) {
+            b_i = i;
+        }
+    }
+
+    return tolower(float_type_list[(a_i > b_i) ? a_i : b_i]);
+}
```

Comment:
Can we do this with `NPY_DOUBLE` instead of `'D'` etc?

Reply:
@eric-wieser Thanks for your review. We can do it but it makes code much complex, I think.
In this function, we compare floating point bit-width including complex types. When we use `NPY_@DTYPE@` enum, we cannot handle the n-bit real type and the n/2-bit complex type as same.

#### Case 1. `NPY_@DTYPE@`
These are not sequential between floating point types (e.g. `NPY_HALF`). Also it's not easy to handle `NPY_@DTYPE@` and `NPY_C@DTYPE@` as same.

#### Case 2. `NPY_@DTYPE@LTR`
An static array cannot hold `toupper(NPY_@DTYPE@LTR)`, and there is no half float complex type. So the `float_type_list[]` becomes
```c
static char float_type_list[] = {'H', NPY_CFLOATLTR, NPY_CDOUBLELTR, NPY_CLONGDOUBLELTR};
```

Reply:
@eric-wieser I got an idea. Simply not using `toupper`.
```c
static char
select_most_suitable_float_type(char a, char b) {
    int i = 0, a_i = 0, b_i = 0;

    static char float_type_list[] = {NPY_HALFLTR, NPY_FLOATLTR, NPY_DOUBLELTR, NPY_LONGDOUBLELTR};

    a = tolower(a);
    b = tolower(b);

    if (!(NPY_DOUBLELTR <= a && a <= NPY_LONGDOUBLELTR)) {
        a = NPY_DOUBLELTR;
    }
    if (!(NPY_DOUBLELTR <= a && a <= NPY_LONGDOUBLELTR)) {
        b = NPY_DOUBLELTR;
    }

    for (i = 0; i < 4; i++) {
        if (a == float_type_list[i]) {
            a_i = i;
        }
        if (b == float_type_list[i]) {
            b_i = i;
        }
    }

    return float_type_list[(a_i > b_i) ? a_i : b_i];
}
```

Reply:
I also wonder if we can reuse the ufunc type resolution machinery here.

Reply:
Your idea sounds interesting. I'll check them out.